### PR TITLE
init InterWasm proposal Oraichain

### PR DESCRIPTION
**Pull request by ducphamle2**

### CODE REVIEW

File: IWPs/iwp-1.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+All of those obstacles have motivated us to build a hub for developers, called CosmWasm IDE. The IDE will provide developers with *a set of quick, convenient, and specific tools to fill the gap between ideas and development, and reduce the time consumption.* With readily available testing and developing environments and a high compatibility level to existing CosmosSDK-based blockchains, developers can easily write code, build, simulate, and deploy their smart contracts.
+
+## Specification
+
+- Use Gitpod template for server-side-development environment. Each user can have access to their workspace with fully customized VS Code extensions, terminals and other tools suitable to build and deploy wasm smart contracts.
+
+- Integrate Keplr wallet extension to deploy the contracts onto different Cosmos-based networks.
+
+- Build a Simulate VS Code extension and install it into the Gitpod template so users can simulate running the contracts as if they have deployed them onto the network.
```
Comment (orkunkl) [inline IWPs/iwp-1.md:28]: What functionality `simulate` have?

Comment (ducphamle2) [reply IWPs/iwp-1.md:28]: @orkunkl Hi, I have just added the ```simulate``` features via a new commit. Also, here's the extension URL on Visual Studio: [Simulate VS extension](https://marketplace.visualstudio.com/items?itemName=oraichain.cosmwasm)

**Comment by Highlander-maker** [comment]
Remix-like Requirements

i am in the process of writing a document based on my experience using Remix in a Solidity env. So i hope this can help with the requirements of what we need to get Rust and CW toolset built. Im super excited about this project. I would love to see it happen. 

UI 

### Left side (Categories)

On the left of the IDE have a browser tab on left where you can create files for your project. It be nice to have it read the languages when you name the file .js .rs ect. 

### Right side (

- Compile

on your right hand side have a ' start to compile ' this compiles the code in to binary form which will be ready for deployment.  if no errors show then you move to the next step. 

- Run

we use this tab to test deploy, remix IDE you have diff env to choose i.e JavaScript VM, Injected Web3 (use Keplr extension) and Web3 Provider for example. This can be a simple deploy button which triggers a script. (Orkun can add to the ideas of ENV here). After execution it shows a nice deployed contract in the ' deployed contracts tab ' here you will find all the get and set functions where you can get value of a variable ect. 

- Settings

 settings tab im not sure what you could add in here. But all n all could be something useful 

- Analysis

  Option to check history of deployed contracts ect. ( More to add) 

- Debugger

  Rust debugger possibly showing here. 

- Support

 Support from CW if things are getting confusing. 

## Deploy Smart Contract

So on this page we need something like the following

- From (connected wallets)

So in order to deploy contract we need to have a page that has our injected web3 wallet. (Keprl) a drop down menu which has FROM that when user wallet is connected it shows our address and balance of wallet. 

Two ways to deploy contract 

- Rust Contract Source Code

Below 'from' drop down possibly have our Rust Source code showing of a particular smart contract code that we create in the previous page of our IDE. 

- Contract Bite Code

Contract bite code to use to deploy also. 

What ever you use it automatically compiles and give you another drop down menu of the type of Smart contracts you would like to deploy. (cw20 ... "my first token" .. mint NFT ect. ) 
Another section i feel can be built solely for Deploy Function capabilities. This section is the next stage along of the users Smart Contract creation. Here we can test the functionality of our code on a selected chain to test before deployment. 

Here are some ideas of what can be in this section

- Deploy function

Hit on the deploy function button and the tx is then completed.

There is another way we can deploy by selecting Injective web3 on our environments tab and with that i should be able to read our address on this IDE which we can deploy contract using it also.

- Wallet 
section showing the balance of a selected wallet. 

- Contracts 
a list of deployed contracts or created tokens on chain 

i will finish the document and send on to Orkin to give it a once over. Also if any more ideas comes to mind as a User i will add to this repo. 

I will dedicate time to this as i think its an awesome idea and would love to see it materialise

File: IWPs/iwp-1.md (hunk 1 above)
Comment (tubackkhoa) [reply IWPs/iwp-1.md:28]: here is the repo of simulate project, the functionalities are described in the README, when integrated with the IDE, it supports hot-reload for fast development.
https://github.com/oraichain/cosmwasm-simulate

**Comment by ducphamle2** [comment]
@Highlander-maker thank you for your comments to improve the CosmWasm IDE. We'll consider your suggestions thoroughly and put them to work. Here are the two public repositories we're working on that you can contribute to if you like: [VS Code CosmWasm](https://github.com/oraichain/vscode-cosmwasm) and [CosmWasm Simulate](https://github.com/oraichain/cosmwasm-simulate)

**Comment by Highlander-maker** [comment]
### After Deployment

So after your contract looks good and you have deployed and ran in your connected enviroment.

after it deploying successfully on the right hand side under RUN section a Deployed Contracts section should be visible to users can interact with their Smart Contract. For example if we have created a Contest smart contract where participants are asked to input some variables to the contract. Name, Phone, Email ect. What if in this section we can add our call functions in there like.... Pick a winner, Registration, transferAmount, getWinnerSelected and getManager. This allows the user to call functions straight off their Project.



Once again if ive not highlighted this before a terminal should be visible on the bottom of the screen too. 

### Create a compile script

```jsx
cosnt path = require('path');
const fs = require('fs'); 

   const solc = require('solc'); 

   const filenamePath = path.resolve(__dirname, 'contracts',      '<filename.rs>';

  const source = fs.readFileSync(filenamePath, 'utf8');

console.log(solc.compile(source, 1));
```

 

once you get the byte code and the interface properties from the compiled code we can add the contract name and export the contract. 

```jsx
/// console.log(solc.compile(source, 1));

module.exports = solc.compile((source, 1).contracts[':name-of-the-contract'];
```

**Comment by ducphamle2** [comment]
> ### After Deployment
> So after your contract looks good and you have deployed and ran in your connected enviroment.
> 
> after it deploying successfully on the right hand side under RUN section a Deployed Contracts section should be visible to users can interact with their Smart Contract. For example if we have created a Contest smart contract where participants are asked to input some variables to the contract. Name, Phone, Email ect. What if in this section we can add our call functions in there like.... Pick a winner, Registration, transferAmount, getWinnerSelected and getManager. This allows the user to call functions straight off their Project.
> 
> 
> 
> Once again if ive not highlighted this before a terminal should be visible on the bottom of the screen too.
> 
> ### Create a compile script
> ```js
> cosnt path = require('path');
> const fs = require('fs'); 
> 
>    const solc = require('solc'); 
> 
>    const filenamePath = path.resolve(__dirname, 'contracts',      '<filename.rs>';
> 
>   const source = fs.readFileSync(filenamePath, 'utf8');
> 
> console.log(solc.compile(source, 1));
> ```
> 
> once you get the byte code and the interface properties from the compiled code we can add the contract name and export the contract.
> 
> ```js
> /// console.log(solc.compile(source, 1));
> 
> module.exports = solc.compile((source, 1).contracts[':name-of-the-contract'];
> ```

Great work you got there. I believe that the GUI part won't be a problem especially when the project is open to the public. The challenging part is to attract more users to create new smart contracts on it. When that time comes, we can scale larger depending on the issues and requests from such users

**Comment by ducphamle2** [comment]
@orkunkl Do you have any additional requirements or requests that I need to do?

**Comment by Highlander-maker** [comment]
I think we will have loads of Devs working on this. With the right marketing i believe we can.. Im glad you like the work and i am buzzing to see it work...

**Comment by orkunkl** [comment]
I am merging the proposal. Let's move the collected info to a new file and launch the project!

**Comment by Highlander-maker** [comment]
Go for it @orkunkl let me know if you need anything done.

**Comment by ducphamle2** [comment]
@orkunkl Here's the Cosmos address that we use to hold the funds: ```cosmos1u54s9xm57r78qcgphx8np0t3kquxywr6pdf679```

**Comment by ducphamle2** [comment]
We have fully received the funds from InterWasm: https://www.mintscan.io/cosmos/txs/949ECA97C94E4571A406EA0ACA28ECFFB7D5120BC4C9E68438E3FF8B79116D35

**Comment by kleimanb81** [comment]
Still active here?

**Comment by ducphamle2** [comment]
@kleimanb81 yep, the project is still in active state

**Comment by kleimanb81** [comment]
Hahah didn't know you were in this, I didn't even remember posting. I was really intrigued. Hows it going?

**Pull request merged**